pkg/cloud/aws/services/ec2/instances.go

```diff
@@ -0,0 +1,305 @@
+	AMI                providerconfigv1.AWSResourceReference
+	SecurityGroups     []providerconfigv1.AWSResourceReference
+	Subnet             providerconfigv1.AWSResourceReference
+	Placement          providerconfigv1.Placement
+	DeviceIndex        int64
+	PublicIP           *bool
+	Tags               []providerconfigv1.TagSpecification
+	IAMInstanceProfile *providerconfigv1.AWSResourceReference
+	InstanceType       string
+	KeyName            *string
+	UserData           []byte
+}
+
+// LaunchInstance launches an instance in AWS
+func (ir *instancesRuntime) LaunchInstance(machine *clusterv1.Machine, lc *LaunchConfig) (*ec2.Instance, error) {
+	// Get AMI to use
+	var amiID *string
+	if lc.AMI.ID != nil {
+		amiID = lc.AMI.ID
+		glog.Infof("Using AMI %s", *amiID)
+	} else if len(lc.AMI.Filters) > 0 {
+		glog.Info("Describing AMI based on filters")
+		describeImagesRequest := ec2.DescribeImagesInput{
+			Filters: buildEC2Filters(lc.AMI.Filters),
+		}
+		describeAMIResult, err := ir.awsclient.DescribeImages(&describeImagesRequest)
+		if err != nil {
+			return nil, fmt.Errorf("error describing AMI: %v", err)
+		}
+		if len(describeAMIResult.Images) < 1 {
+			return nil, fmt.Errorf("no image for given filters not found")
+		}
+		latestImage := describeAMIResult.Images[0]
+		latestTime, err := time.Parse(time.RFC3339, *latestImage.CreationDate)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse time for %q AMI: %v", *latestImage.ImageId, err)
+		}
+		for _, image := range describeAMIResult.Images[1:] {
+			imageTime, err := time.Parse(time.RFC3339, *image.CreationDate)
+			if err != nil {
+				return nil, fmt.Errorf("unable to parse time for %q AMI: %v", *image.ImageId, err)
+			}
+			if latestTime.Before(imageTime) {
+				latestImage = image
+				latestTime = imageTime
+			}
+		}
+		amiID = latestImage.ImageId
+	} else {
+		return nil, fmt.Errorf("AMI ID or AMI filters need to be specified")
+	}
+
+	securityGroupsIDs, err := ir.getSecurityGroupsIDs(lc.SecurityGroups)
+	if err != nil {
+		return nil, fmt.Errorf("error getting security groups IDs: %v,", err)
+	}
+	subnetIDs, err := ir.getSubnetIDs(lc.Subnet, lc.Placement.AvailabilityZone)
+	if err != nil {
+		return nil, fmt.Errorf("error getting subnet IDs: %v,", err)
+	}
+	if len(subnetIDs) > 1 {
+		glog.Warningf("More than one subnet id returned, only first one will be used")
+	}
+
+	// build list of networkInterfaces (just 1 for now)
+	var networkInterfaces = []*ec2.InstanceNetworkInterfaceSpecification{
+		{
+			DeviceIndex:              aws.Int64(lc.DeviceIndex),
+			AssociatePublicIpAddress: lc.PublicIP,
+			SubnetId:                 subnetIDs[0],
+			Groups:                   securityGroupsIDs,
+		},
+	}
+
+	// Add tags to the created machine
+	rawTagList := []*ec2.Tag{}
+	for _, tag := range lc.Tags {
+		rawTagList = append(rawTagList, &ec2.Tag{Key: aws.String(tag.Name), Value: aws.String(tag.Value)})
+	}
+
+	clusterID, ok := machine.Labels[providerconfigv1.ClusterIDLabel]
```

Comment:
where did getClusterID go?

Reply:
Dropped. To few information in a single function. Given we are getting away from the cluster object and putting the cluster ID into annotation is the only way how to signify the cluster, we don't need to abstract on the cluster retrieval.